File: Algorithms/746. Min Cost Climbing Stairs/golang/solution.memoization.go
func minCostClimbingStairs(cost []int) int {
	var memo map[int]int = make(map[int]int)
	return minimumCost(cost, memo, len(cost))
}

func minimumCost(cost []int, memo map[int]int, i int) int {
	if i <= 1 {
		return 0
	}

	if val, ok := memo[i]; ok {
		return val
	}

	var downOne int = minimumCost(cost, memo, i-1) + cost[i-1]
	var downTwo int = minimumCost(cost, memo, i-2) + cost[i-2]
	if downOne > downTwo {
		memo[i] = downTwo
	} else {
		memo[i] = downOne
	}
	return memo[i]
}

File: Algorithms/746. Min Cost Climbing Stairs/swift/solution.memoization.swift
class Solution {
    func minCostClimbingStairs(_ cost: [Int]) -> Int {
        func minimumCost(_ i: Int) -> Int {
            guard i > 1 else {
                return 0
            }

            if let val = memo[i] {
                return val
            }

            let downOne: Int = minimumCost(i - 1) + cost[i - 1]
            let downTwo: Int = minimumCost(i - 2) + cost[i - 2]
            memo[i] = min(downOne, downTwo)
            return memo[i]!
        }

        var memo = [Int: Int]()
        return minimumCost(cost.count)
    }
}
